### ISSUE TITLE
Filter Query on a ManagedSet

### ISSUE THREAD

**Issue by KRIOFT** [stablekernel/aqueduct#164]
I want to filter a "city" ManagedObject by it's "starts" ManagedSet.
```
var citiesQuery = new Query<City>()
      ..matchOn.starts.includeInResultSet = true
      ..matchOn.starts.isNotEmpty;
return new Response.ok(await citiesQuery.fetch());
  ```
But this code return all City records from database. Output is this.
```
[
  {
    "id": 1,
    "name": "București",
    "coordinates": "44.4377401,25.9545541",
    "country": {
      "id": 1
    },
    "starts": [
      {
        "id": 1,
        "start": {
          "id": 1
        },
        "end": {
          "id": 2
        }
      }
    ]
  },
  {
    "id": 2,
    "name": "Londra",
    "coordinates": "51.528308,-0.3817765",
    "country": {
      "id": 2
    },
    "starts": []
  },
  {
    "id": 3,
    "name": "Berlin",
    "coordinates": "52.5335457,13.5169484",
    "country": {
      "id": 3
    },
    "starts": []
  }
]
```
Please help me with this situation.

**Comment by joeconwaystk** [comment]
That's a good request. It's not currently possible, but it should be with (what I think) is a minor change and a slightly different syntax.

Currently, the query builder works by assigning properties in `matchOn` to 'matchers' (values/methods that start with the word `where`). So, adding isEmpty to the matcher wouldn't alter the query, but there should be a way to filter the result set in this way. I'll investigate a level of effort for making this work, as it should be a relatively common use case. The syntax will likely be something like this:

```
var citiesQuery = new Query<City>()
      ..matchOn.starts.includeInResultSet = true
      ..matchOn.starts = whereNotNull;
```

**Comment by KRIOFT** [comment]
Ok. I wait a solution for that, beacause a have a project who needs this filter. Thanks.

**Comment by KRIOFT** [comment]
I try to find a alternative solution like:
``` 
class City extends ManagedObject<_City> implements _City {
  int get startsnumber => startsnumber;
  void set startsnumber(un) {
    starts.length = un;
  }
}

class _City {
  @managedPrimaryKey
  int id;

  @ManagedColumnAttributes(indexed: true)
  String name;

  @ManagedColumnAttributes(unique: true)
  String coordinates;

  @ManagedRelationship(#cities)
  Country country;

  ManagedSet<Route> starts;
  ManagedSet<Route> ends;
}
```
And I get ``` NoSuchMethodError: Class 'ComparisonMatcherExpression' has no instance method '>'.```

The code for controller is: 
```
@httpGet
  getStartCities() async {
    var citiesQuery = new Query<City>()
      ..matchOn.starts.includeInResultSet = true
      ..matchOn.startsnumber = whereGreaterThan(0);

    return new Response.ok(await citiesQuery.fetch());
  }
```

**Comment by joeconwaystk** [comment]
The properties declared on City aren't database columns, so you won't be able to include those in a query. A workaround right now would either be fetching all of the cities and then filtering them in the application:

```dart
var q = new Query<City>()
  ..matchOn.starts.includeInResultSet = true;

var cities = (await q.fetch())
  .where((city) => city.starts != null)
  .toList();
```

Or it would be to do two queries of questionable performance:

```dart
var q = new Query<Route>()
  ..matchOn.cityStart = whereNotNull;
var allRoutes = await q.fetch();

var cityQuery = new Query<City>()
  ..matchOn.id = whereIn(allRoutes.map((r) => r.cityStart.id));
var cities = await cityQuery.fetch();

// pair together cities with their routes.
```

**Comment by KRIOFT** [comment]
When you send me this message, I was trying first option :)). And it works.
I write more code, so I am able to do something wrong :)).

So please put in todo list :
- filter empty for 'has one' and 'has many' relations
- get in transident managed object attribute the count of rows of some ManagedSet from persistent type

P.S. I like Aqueduct and I want to use it in production. I am a novice in Dart and your project, but I was tried to look in source and in this moment is too much for me. I will try to contribuite in the future.

**Comment by joeconwaystk** [comment]
This is now resolved for 2.0
